### PR TITLE
Adds a feature to import delegate records during PICA import

### PR DESCRIPTION
**Pull request by matthias-ronge**
By an additional entry in opac.xml, a “resolve rule” can be defined. If a field in the imported PICA hit matches the rule’s `tag` and `subtag`, the value from that subtag is used to query the catalogue again. In a mapping, values from the result of that second query can be added as (fictious) subfields to the title field that the query was started from.

Example mapping:

    <catalogue title="…">
        
        <resolve tag="028C" subtag="9" searchField="12">
            <map tag="003U" subtag="a" asSubtag="g" />
        </resolve>
    
Example field on a title record:
    
    <field tag="028C">
        <subfield code="d">Dirk</subfield>
        <subfield code="a">Fahlenkamp</subfield>
        <subfield code="9">69749487X</subfield>
        <subfield code="8">Fahlenkamp, Dirk *1952-*</subfield>
    </field>
    
Example result after successful resolving:
    
    <field tag="028C">
        <subfield code="d">Dirk</subfield>
        <subfield code="a">Fahlenkamp</subfield>
        <subfield code="9">69749487X</subfield>
        <subfield code="8">Fahlenkamp, Dirk *1952-*</subfield>
        <subfield code="g">http://d-nb.info/gnd/172559227</subfield>
    </field>

This is part of issue #619.

**Additional edit:** German javadoc comments in `PicaPlugin.java` have been translated into English.

### CODE REVIEW

**Comment by matthias-ronge** [comment]
Note: I could not yet test this implementation agains a real PICA catalogue because GBV catalogue does not expose norm data records to the public. I tested it with a local replacement instead, where I could prove that the function as-is does work. We are working together with Staatsbibliothek zu Berlin − Preußischer Kulturbesitz to get a real test environment, but I wanted to publish the code already, so that you can review it when it fits into your schedule.

File: Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/CatalogueClient.java [hunk 1]
```diff
@@ -448,4 +513,8 @@ private Response parseOpacResponse(String opacResponse) throws IOException, SAXE
 
 		return ids;
 	}
+
+	public void setTimeout(long timeout) {
```
Comment (henning-gerhardt) [inline Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/CatalogueClient.java]: missing JavaDoc statement

File: Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/PicaPlugin.java [hunk 2]
```diff
@@ -234,19 +236,19 @@ public static String getDescription(Locale language) {
 		Type type;
 		Fileformat ff;
 		try {
-			/*
-			 * -------------------------------- Opac abfragen und erhaltenes
-			 * Dom-Dokument in JDom-Dokument umwandeln
+			/* 
+			 * --------------------------------
```
Comment (henning-gerhardt) [inline Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/PicaPlugin.java]: General hint / question for whole pull request: On cleaning up and translations comments, please remove this unnecessary `---`,  `===` or `###` lines. One line comments should started with `//`, two or more line comments should / could be surrounded in `/*` ... `*/`

File: Goobi/config/goobi_opac.xml [hunk 3]
```diff
@@ -39,6 +39,9 @@
 			<searchField label="Barcode" value="8535" />
 			<searchField label="Barcode 8200" value="8200" />
 		</searchFields>
+		<resolve tag="028C" subtag="9">
```
Comment (henning-gerhardt) [inline Goobi/config/goobi_opac.xml]: Additional comment: Your pull request example add a `searchField` attribute but this is missing or this attribute optional?

File: Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/Catalogue.java [hunk 4]
```diff
@@ -281,4 +283,8 @@ String getUncf() {
 		return ucnf;
 	}
 
+	Map<String,ResolveRule> getResolveRules() {
```
Comment (matthias-ronge) [inline Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/Catalogue.java]: Should there be javadoc statements for functions that are package private?

Comment (henning-gerhardt) [inline Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/Catalogue.java]: Methods without a `private`, `protected` or `public` I read them every time as `public` :( Add a Javadoc comment if needed (method to complex, made more things as the name suggest, ...).

File: Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/MapRule.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+package org.goobi.production.plugin.CataloguePlugin.PicaPlugin;
+
+import org.apache.commons.configuration.HierarchicalConfiguration;
+
+/**
+ * Map rules are used to map data from norm data records to a media record. This
+ * is used inside a {@code ResolveRule}. Example usage:
+ *
+ * <pre>
+ * &lt;catalogue title="…">
```
Comment (matthias-ronge) [inline Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/MapRule.java]: Yes, readability. Compare XML encoding:
```
&lt;catalogue title=&quot;&#8230;&quot;&gt;
    &lt;!-- ... --&gt;
    &lt;resolve tag=&quot;028C&quot; subtag=&quot;9&quot; searchField=&quot;12&quot;&gt;
        &lt;map tag=&quot;003U&quot; subtag=&quot;a&quot; asSubtag=&quot;g&quot; /&gt;
    &lt;/resolve&gt;
```
See also [this answer](http://stackoverflow.com/a/10462413/1503237).

Comment (henning-gerhardt) [inline Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/MapRule.java]: It is more readable then a XML encoding but even still horrible in my eyes. If I look into [Sun/Oracle documentation](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html) and [complex examples](http://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#examples) I don't see there any of encoding tags. Maybe this is a some strange behaviour of some IDE's which made this kind of documentation non-readable for plain-text users.

Comment (henning-gerhardt) [inline Goobi/plugins/opac/PicaPlugin/org/goobi/production/plugin/CataloguePlugin/PicaPlugin/MapRule.java]: If there is no other way, we should tolerate this.

File: Goobi/config/goobi_opac.xml (hunk 3 above)
Comment (matthias-ronge) [inline Goobi/config/goobi_opac.xml]: Attributes `searchField` and `substring` are optional, but should be documented in the WIKI.

**Pull request merged**